modal/runner/shared/download.py

```diff
@@ -15,47 +16,42 @@
 )
 
 
-def download_models(all_models: List[str]):
-    from os import environ as env
-
+@stub.function(
+    image=downloader_image,
+    volumes={models_path: models_volume},
+    secret=Secret.from_name("huggingface"),
+    timeout=3600,  # 1 hour per model
+)
+def download_model(model_name: str):
     from huggingface_hub import list_repo_files, snapshot_download
 
-    cache_path = get_model_path("__cache__")
-    for model_name in all_models:
-        model_path = get_model_path(model_name)
-        model_path.mkdir(parents=True, exist_ok=True)
-
-        try:
-            # only download safetensors if available
-            has_safetensors = any(
-                fn.lower().endswith(".safetensors")
-                for fn in list_repo_files(
-                    model_name,
-                    repo_type="model",
-                    token=env["HUGGINGFACE_TOKEN"],
-                )
-            )
-            patterns = ["tokenizer.model", "*.json"]
-            if has_safetensors:
-                patterns.append("*.safetensors")
-            else:
-                patterns.append("*.bin")
-
-            snapshot_download(
-                model_name,
-                local_dir=model_path,
-                local_files_only=True,
-                cache_dir=cache_path,
-                token=env["HUGGINGFACE_TOKEN"],
-            )
-            print(f"Volume contains {model_name}")
-        except FileNotFoundError:
-            print(f"Downloading {model_name} ...")
-            snapshot_download(
-                model_name,
-                local_dir=model_path,
-                cache_dir=cache_path,
-                token=env["HUGGINGFACE_TOKEN"],
-            )
-            stub.models_volume.commit()
-    print("ALL DONE!")
+    model_path = get_model_path(model_name)
+    model_path.mkdir(parents=True, exist_ok=True)
+
+    # only download safetensors if available
+    has_safetensors = any(
+        fn.lower().endswith(".safetensors")
+        for fn in list_repo_files(
+            model_name,
+            repo_type="model",
+            token=env["HUGGINGFACE_TOKEN"],
+        )
+    )
+    patterns = ["tokenizer.model", "*.json"]
+    if has_safetensors:
+        patterns.append("*.safetensors")
+    else:
+        patterns.append("*.bin")
+    # TODO: Use these patterns?
+
+    # Clean doesn't remove the cache, so using `local_files_only` here returns the cache even when the local dir is empty.
+    print(f"Checking for {model_name}")
+    snapshot_download(
+        model_name,
+        local_dir=model_path,
+        cache_dir=cache_path,
+        local_dir_use_symlinks=False,
```

Comment:
I think we should use sym_links here to download the file once to cache only :-?... - and maybe we should fix the clean method to purge cache instead?